scripts: code_relocate: support section filter

schouleu opened the pull request:
One might want to select the symbols to be relocated inside a file or a library.
To do this, one can use the FILTER argument of zephyr_code_relocate which 
must contain a regular expression of the section names to be selected for relocation.

This pull-request covers the issue https://github.com/zephyrproject-rtos/zephyr/issues/64148

About the dependency on -ffunction-sections and -fdata-sections flags, those are enabled by default in [arch/common//home/schouleur/GML/zephyr_upstream/zephyr/arch/common/CMakeLists.txt:37](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/common/CMakeLists.txt#L37)

In case the flags are not enabled, this feature can also be used to select some specific sections.

schouleu committed:
scripts: code_relocate: support section filter

One might want to select the symbols to be relocated inside a file or
a library. To do this, one can use the FILTER argument of
zephyr_code_relocate which must contain a regular expression of the
section names to be selected for relocation.

The test_function_in_sram2 test case in
`tests/application_development/code_relocation` has been updated to
verify that only one function `function_in_sram()` is relocated to ram
and that the function `function_not_relocated()` is not being relocated
when using relocation filter.

Signed-off-by: Sylvain Chouleur <sylvain.chouleur@gmail.com>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented on scripts/build/gen_relocate_app.py:
```diff
@@ -529,11 +536,16 @@ def create_dict_wrt_mem():
     if input_rel_dict == '':
         sys.exit("Disable CONFIG_CODE_DATA_RELOCATION if no file needs relocation")
+
+    # Replace pipes used in symbol filter regular expression
+    input_rel_dict = input_rel_dict.replace("\\|", "##PIPE##")
```
I know the `|` is used widely in regex'es, so it's natural to support them directly, but going through this substitution seems less than optimal to me.

This PR https://github.com/zephyrproject-rtos/zephyr/pull/60999 changed the input relocation argument from being a string to be a file.

Being a file actually means that there is no reason for a single string because a file can be multi-line.
I suggest to make a #60999 follow-up commit which remove the use of `|` as separator and instead create a multi-line input file.

Doing so will free the use of `|` and thereby remove the need for special handling in the regex.

tejlmand commented on scripts/build/gen_relocate_app.py:
```diff
@@ -514,12 +520,13 @@ def parse_input_string(line):
 
     # Split lists by semicolons, in part to support generator expressions
-    flag_list, file_list = (lst.split(';') for lst in rest.split(':', 1))
+    flag_list, file_list, symbol_filter = (lst.split(';') for lst in rest.split(':', 2))
```
This fails on windows, most likely because drive letters contains `:`.

```
c:/users/torsten/zephyr-sdk-0.16.8/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe: app/libapp.a(test_file1.c.obj): in function `code_relocation_test_function_in_sram2':
C:/projects/github/ncs/zephyr/tests/application_development/code_relocation/src/test_file1.c:80: undefined reference to `__sram2_data_reloc_start'
c:/users/torsten/zephyr-sdk-0.16.8/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe: C:/projects/github/ncs/zephyr/tests/application_development/code_relocation/src/test_file1.c:80: undefined reference to `__sram2_data_reloc_end'
....
```

Also note the existing warning in this script:
https://github.com/zephyrproject-rtos/zephyr/blob/878640fa0817a5c6684758d68dcb824e24571612/scripts/build/gen_relocate_app.py#L508

schouleu replied:
@tejlmand do you have a way to test on windows without windows?

tejlmand replied:
unfortunately not.

carlescufi commented:
@schouleu do you have plans to address @tejlmand's comments?

schouleu commented:
Hi @carlescufi 
Yes, I don't have a lot of time right now but I'll implement @tejlmand's suggestions

carlescufi commented:
> Hi @carlescufi Yes, I don't have a lot of time right now but I'll implement @tejlmand's suggestions

@schouleu it's been a while, would you mind if we take over the PR?

carlescufi commented:
@schouleu would you mind if we take over your PR as per the process described here:
https://docs.zephyrproject.org/latest/contribute/modifying_contributions.html

schouleu commented:
> @schouleu would you mind if we take over your PR as per the process described here: https://docs.zephyrproject.org/latest/contribute/modifying_contributions.html

Sure, go ahead

tejlmand committed:
cmake: code relocation setting.

With code relocation directives passed to the gen_relocate_app.py script
using generated file, then each directive can be place on individual
line in the file and thus free up the `|` character as separator.

Furthermore, a multi-line file with each directive on separate line is
also more user-readable, making debugging easier.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> > @schouleu would you mind if we take over your PR as per the process described here: https://docs.zephyrproject.org/latest/contribute/modifying_contributions.html
> 
> Sure, go ahead

PR rebased and updated in following areas:
- Follow-up commit to #60999 added which removes `|` as directive separator and instead places each directive on independent line, as #60999 introduced the use of input files, see comment: https://github.com/zephyrproject-rtos/zephyr/pull/74402#discussion_r1715401550
- Changed separator between files and filter argument to `,`, so that it doesn't collide with windows drive letter symbol, `:`.

57300 commented on scripts/build/gen_relocate_app.py:
```diff
@@ -243,6 +245,9 @@ def find_sections(filename: str) -> 'dict[SectionKind, list[OutputSection]]':
         out = defaultdict(list)
 
         for section in sections:
+            if len(symbol_filter) > 0 and not re.search(symbol_filter, section.name):
```
nit: The first condition should not be needed, as long as `symbol_filter` is an actual string.
```suggestion
            if not re.search(symbol_filter, section.name):
```

tejlmand replied:
Fixed.

57300 commented on scripts/build/gen_relocate_app.py:
```diff
@@ -513,27 +519,31 @@ def parse_input_string(line):
+
     # Split lists by semicolons, in part to support generator expressions
-    flag_list, file_list = (lst.split(';') for lst in rest.split(':', 1))
+    file_list, symbol_filter = (lst.split(';') for lst in rest.split(',', 1))
```
Why is `symbol_filter` split by `;` and returned as a list?

tejlmand replied:
it's because that allows the use of generator expressions from CMake which could form a `;`-separated list.
Used already today for file lists, and considered it would make sense to support the same for the symbol filter, but can take it out if prefered.

Can be added when there is an actual need.

57300 replied:
Okay. Currently, only the first regex seems to be used. If we were to allow multiple, we'd need to decide whether to treat as AND- or OR-clauses, but all of this can be described in a single regex. A CMake genex can also be joined into one.

tejlmand replied:
Removed.

57300 commented on scripts/build/gen_relocate_app.py:
```diff
@@ -513,27 +519,31 @@ def parse_input_string(line):
+    file_list, symbol_filter = (lst.split(';') for lst in rest.split(',', 1))
 
-    return mem_region, phdr, flag_list, file_list
+    return mem_region, phdr, [flag_list], file_list, symbol_filter
```
`flag_list` needs to be split by `;` before returning. This used to happen in the previous line.

tejlmand replied:
Good point, will fix.

57300 commented on scripts/build/gen_relocate_app.py:
```diff
@@ -513,27 +519,31 @@ def parse_input_string(line):
 
-    input_rel_dict = args.input_rel_dict.read()
+    input_rel_dict = args.input_rel_dict.read().splitlines()
     if input_rel_dict == '':
```
```suggestion
    if not input_rel_dict:
```
`input_rel_dict` is now a list.

tejlmand replied:
fixed

nordicjm replied:
think the fix for this is in wrong commit

tejlmand replied:
thanks for spotting, fixed.

57300 commented on scripts/build/gen_relocate_app.py:
```diff
@@ -556,12 +566,15 @@ def create_dict_wrt_mem():
             print("Memory region ", mem_region, " Selected for files:", file_name_list)
 
+        # Apply filter on files
+        file_name_filter_list = [(f, symbol_filter[0]) for f in file_name_list]
```
I don't think `symbol_filter` should be a list.

tejlmand replied:
I don't really have an opinion on that.
My main focus was to address the `:` issue (windows support) and `|` separator usage.

Can remove it, but would like to hear the intention from original author, @schouleu ?

schouleu replied:
The symbol_filter is supposed to be a single string regexp, I don't see any need to have a list of filters since the regexp itself can handle multiple matches

tejlmand replied:
Removed.

57300 reviewed (changes requested):
Thanks for taking this on, @schouleu and @tejlmand

@tejlmand, could you add a few tests to `tests/application_development/code_relocation`?

I haven't tested this PR iteration, but I'm noticing some oddities in the code.

tejlmand commented:
> @tejlmand, could you add a few tests to `tests/application_development/code_relocation`?

Extended the `file2.c` containing `function_in_sram()` to also include `function_not_relocated()`, and made a verification that `function_in_sram()` is actually being relocated to ram.
I discovered that current test on zephyr/main is actually not verifying the location of the function, meaning that even in the case that he function was not properly relocated then the test would still pass :open_mouth: .

This is now fixed, together with the new, non-relocated, function.
